lib/os/onoff.c

```diff
@@ -0,0 +1,627 @@
+
+	switch (state) {
+	case SERVICE_STATE_ON:
+		/* Stay on if release leaves a client. */
+		if (srv->refs > 1U) {
+			notify = true;
+			rv = 1;
+			break;
+		}
+
+		/* Reject if in no-wait context but stop could
+		 * wait.
+		 */
+		if ((no_wait || k_is_in_isr())
+		    && ((srv->flags & ONOFF_SERVICE_STOP_WAITS) != 0)) {
+			rv = -EWOULDBLOCK;
+			break;
+		}
+
+		stop = true;
+
+		set_service_state(srv, SERVICE_STATE_TO_OFF);
+		srv->releaser = cli;
+		rv = 2;
+
+		break;
+	case SERVICE_STATE_TO_ON:
+		rv = -EBUSY;
+		break;
+	case SERVICE_STATE_OFF:
+	case SERVICE_STATE_TO_OFF:
+		rv = -EALREADY;
+		break;
+	default:
+		rv = -EINVAL;
+	}
+
+out:
+	if (notify) {
+		srv->refs -= 1U;
+	}
+
+	k_spin_unlock(&srv->lock, key);
+
+	if (stop) {
+		__ASSERT_NO_MSG(srv->stop != NULL);
+		srv->stop(srv, onoff_stop_notify);
+	} else if (notify) {
+		notify_one(srv, cli, 0);
+	}
+
+	return rv;
+}
+
+static void onoff_reset_notify(struct onoff_service *srv,
+			       int res)
+{
+	k_spinlock_key_t key = k_spin_lock(&srv->lock);
+	sys_slist_t clients = srv->clients;
+
+	/* If the reset failed clear the transition flag but otherwise
+	 * leave the state unchanged.
+	 *
+	 * If it was successful clear the reference count and all
+	 * flags except capability flags.
+	 */
+	if (res < 0) {
+		srv->flags &= ~SERVICE_STATE_TRANSITION;
+	} else {
+		__ASSERT_NO_MSG(srv->refs == 0U);
+		srv->refs = 0U;
+		srv->flags &= SERVICE_CONFIG_FLAGS;
+	}
+
+	sys_slist_init(&srv->clients);
+
+	k_spin_unlock(&srv->lock, key);
+
+	notify_all(srv, &clients, res);
+}
+
+int z_impl_onoff_service_reset(struct onoff_service *srv,
+			       struct onoff_client *cli)
+{
+	if (srv->reset == NULL) {
+		return -ENOTSUP;
+	}
+
+	bool no_wait = false;
+	bool reset = false;
+	int rv = validate_args(srv, cli, &no_wait);
+
+	if (rv < 0) {
+		return rv;
+	}
+
+	if (rv == ONOFF_CLIENT_NOTIFY_ANONYMOUS) {
+		return -ENOTSUP;
+	}
+
+	rv = 0;
+
+	/* Reject if in a no-wait context and reset could
+	 * wait.
+	 */
+	if ((no_wait || k_is_in_isr())
+	    && ((srv->flags & ONOFF_SERVICE_RESET_WAITS) != 0U)) {
+		return -EWOULDBLOCK;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&srv->lock);
+
+	if ((srv->flags & ONOFF_SERVICE_HAS_ERROR) == 0) {
+		rv = -EINVAL;
+		goto out;
+	}
+
+	if ((srv->flags & SERVICE_STATE_TRANSITION) == 0) {
+		reset = true;
+		srv->flags |= SERVICE_STATE_TRANSITION;
+	}
+
+out:
+	if (rv >= 0) {
+		sys_slist_append(&srv->clients, &cli->node);
+	}
+
+	k_spin_unlock(&srv->lock, key);
+
+	if (reset) {
+		srv->reset(srv, onoff_reset_notify);
+	}
+
+	return rv;
+}
+
+int z_impl_onoff_cancel(struct onoff_service *srv,
+			struct onoff_client *cli)
+{
+	bool no_wait = false;
+	int rv = validate_args(srv, cli, &no_wait);
+
+	if (rv < 0) {
+		return rv;
+	}
+
+	if (rv == ONOFF_CLIENT_NOTIFY_ANONYMOUS) {
+		return -ENOTSUP;
+	}
+
+	rv = -EALREADY;
+	k_spinlock_key_t key = k_spin_lock(&srv->lock);
+
+	if (sys_slist_find_and_remove(&srv->clients, &cli->node)) {
```

Comment:
any reason why this cannot be called here https://github.com/zephyrproject-rtos/zephyr/blob/577f2ac8374a5bf6386345d534c1664ca18838b3/lib/os/onoff.c#L472 ?

Reply:
If the service has not started, and the client is attempting to release it, the client is supposed to cancel the request instead.

This is the sort of slippery slope that gets started when you allow anonymous requests.  Between this, and the request to forget about a releaser, I'm going to remove that "feature": it's too hard to be sure all cases are handled.